core/include/traccc/fitting/kalman_filter/kalman_fitter.hpp

```diff
@@ -0,0 +1,201 @@
+                     fit_actor_state, m_resetter_state);
+
+        // Create propagator state
+        typename propagator_type::state propagation(
+            seed_params, m_detector->get_bfield(), *m_detector);
+
+        // Set overstep tolerance and stepper constraint
+        propagation._stepping().set_overstep_tolerance(
+            m_cfg.overstep_tolerance);
+        propagation._stepping
+            .template set_constraint<detray::step::constraint::e_accuracy>(
+                m_cfg.step_constraint);
+
+        // Run forward filtering
+        propagator.propagate(propagation, actor_states);
+
+        // Run smoothing
+        smooth(fit_actor_state.m_track_states);
+
+        //@todo: Write track info
+    }
+
+    /// Run smoothing after kalman filtering
+    ///
+    /// @brief The smoother is based on "Application of Kalman filtering to
+    /// track and vertex fitting", R.Frühwirth, NIM A.
+    ///
+    /// @param track_states the vector of track state
+    template <typename track_state_collection_t>
+    void smooth(track_state_collection_t& track_states) {
+
+        // The smoothing algorithm requires the following:
+        // (1) the filtered track parameter of the current surface
+        // (2) the smoothed track parameter of the next surface
+        //
+        // Since the smoothed track parameter of the last surface can be
+        // considered to be the filtered one, we can reversly iterate the
+        // algorithm to obtain the smoothed parameter of other surfaces
+        auto& last = track_states.back();
+        last.smoothed().set_vector(last.filtered().vector());
+        last.smoothed().set_covariance(last.filtered().covariance());
+
+        const auto& mask_store = m_detector.get()->mask_store();
+
+        for (typename track_state_collection_t::reverse_iterator it =
+                 track_states.rbegin() + 1;
+             it != track_states.rend(); ++it) {
+
+            // Surface
+            const auto& surface =
+                m_detector.get()->surface_by_index(it->surface_link());
+
+            // Run kalman smoother
+            mask_store.template call<gain_matrix_smoother<transform3_type>>(
+                surface.mask(), *it, *(it - 1));
+        }
+    }
+
+    /// @return fitter result of a track
+    fitter_info<transform3_type> get_fitter_info() const { return m_fit_info; }
+
+    /// @return fitter result of each track states
+    vecmem::vector<track_state<transform3_type>> get_track_states() const {
+        return m_track_states;
+    }
+
+    private:
+    std::unique_ptr<detector_type> m_detector;
+    fitter_info<transform3_type> m_fit_info;
+    vecmem::vector<track_state<transform3_type>> m_track_states;
+
+    typename aborter::state m_aborter_state{};
+    typename transporter::state m_transporter_state{};
+    typename interactor::state m_interactor_state{};
+    typename resetter::state m_resetter_state{};
```

Comment:
I don't like this. :frowning: We should not have a state on the fitter object.

Any "state" that is needed for the fitting should be put into its own object, which the client of the fitter would manage in memory. So that multiple clients could call on the same fitter object, each with their own "state objects".

I.e. you should introduce something like a `traccc::kalman_fitter::state` struct, which would have all these variables. And then all the clients would need to pass instances of that state type to the `fit(...)`, `filter(...)` and `smooth(...)` functions.

Reply:
I made the changes based on suggestion.
Please note that the kalman fitter state and kalman actor state will be updated again when I implement the GPU kalman fitting.